src/Microsoft.ML.OnnxTransform/DnnImageFeaturizerTransform.cs

```diff
@@ -0,0 +1,116 @@
+        /// just verifies the schema.
+        /// </summary>
+        public TransformerChain<OnnxTransform> Fit(IDataView input)
+        {
+            return _modelChain.Fit(input);
+        }
+
+        public SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            return _modelChain.GetOutputSchema(inputSchema);
+        }
+    }
+
+    public static class DnnImageFeaturizerStaticExtensions
+    {
+        private sealed class OutColumn : Vector<float>
+        {
+            public PipelineColumn Input { get; }
+
+            public OutColumn(Vector<float> input, Func<DnnImageModelSelector, IHostEnvironment, string, string, EstimatorChain<OnnxTransform>> modelFactory)
+                : base(new Reconciler(modelFactory), input)
+            {
+                Input = input;
+            }
+        }
+
+        private sealed class Reconciler : EstimatorReconciler
+        {
+            private readonly Func<DnnImageModelSelector, IHostEnvironment, string, string, EstimatorChain<OnnxTransform>> _modelFactory;
+
+            public Reconciler(Func<DnnImageModelSelector, IHostEnvironment, string, string, EstimatorChain<OnnxTransform>> modelFactory)
+            {
+                _modelFactory = modelFactory;
+            }
+
+            public override IEstimator<ITransformer> Reconcile(IHostEnvironment env,
+                PipelineColumn[] toOutput,
+                IReadOnlyDictionary<PipelineColumn, string> inputNames,
+                IReadOnlyDictionary<PipelineColumn, string> outputNames,
+                IReadOnlyCollection<string> usedNames)
+            {
+                Contracts.Assert(toOutput.Length == 1);
+
+                var outCol = (OutColumn)toOutput[0];
+                return new DnnImageFeaturizerEstimator(env, _modelFactory, inputNames[outCol.Input], outputNames[outCol]);
+            }
+        }
+
+        /// <summary>
+        /// Creates and applies a DnnImageFeaturizer transform to be used by the static API.
+        /// <see cref="DnnImageFeaturizerEstimator"/> for more information about how the transformation works.
+        /// </summary>
+        /// <param name="input">Vector of image pixel weights.</param>
+        /// <param name="modelFactory">An extension method on the <see cref="DnnImageModelSelector"/> that creates a chain of two
+        /// <see cref="OnnxScoringEstimator"/>s (one for preprocessing and one with a pretrained image DNN) with specific models
+        /// included in a package together with that extension method.
+        /// For an example, see Microsoft.ML.DnnImageFeaturizer.ResNet18 </param>
+        /// <returns>A vector of float feature weights based on the input image.</returns>
+        public static Vector<float> DnnImageFeaturizer(this Vector<float> input, Func<DnnImageModelSelector, IHostEnvironment, string, string, EstimatorChain<OnnxTransform>> modelFactory)
```

Comment:
I don't really like this `Func` usage in our public API. What are the two strings? If we had an object representing the inputs to this factory, it would be more self-documenting. And also we could add more optional input parameters in the future.

I think we should come up with a better API than this. My suggestion would be something along the lines of:

```C#
public class DnnImageFeaturizerInput
{
 // all the properties being passed to the factory
}

...

public static Vector<float> DnnImageFeaturizer(this Vector<float> input, Func<DnnImageFeaturizerInput, EstimatorChain<OnnxTransform>> modelFactory);
```

Reply:
Reworked it to use a separate object as suggested. Agree that it makes it a lot more understandable, thanks for the suggestion!

---
In reply to: [232770657](https://github.com/dotnet/machinelearning/pull/1447#discussion_r232770657) [](ancestors = 232770657)